### PR TITLE
Add power_outage_memory and switch_type to Aqara SSM-U01

### PR DESCRIPTION
**Pull request by dmitrydyugaev**
Add support of two features to Aqara SSM-U01.
power_outage_memory - enable/disable the power outage memory
switchtype - toggle or momentary

As it is the first commit to z2m, please guide me in the right direction if I did something wrong.

### CODE REVIEW

File: converters/fromZigbee.js [hunk 1]
```diff
@@ -4370,6 +4370,26 @@ const converters = {
...
+        cluster: 'aqaraOpple',
+        type: ['attributeReport', 'readResponse'],
+        convert: (model, msg, publish, options, meta) => {
+            if (msg.data.hasOwnProperty(0x000A)) {
+                const lookup = {1: 'toggle', 2: 'momentary'};
+                return {switchtype: lookup[msg.data[0x000A]]};
+            }
+        },
+    },
+    aqara_power_outage_memory: {
```
Comment (Koenkk) [inline converters/fromZigbee.js:4383]: Please rename this to `xiaomi_switch_power_outage_memory` and add this to all devices using `tz.xiaomi_switch_power_outage_memory`

File: converters/fromZigbee.js [hunk 2]
```diff
@@ -4370,6 +4370,26 @@ const converters = {
             return result;
         },
     },
+    aqara_switchtype: {
```
Comment (Koenkk) [inline converters/fromZigbee.js:4373]: Please rename this to `xiaomi_switch_type`.

File: converters/fromZigbee.js [hunk 3]
```diff
@@ -4370,6 +4370,26 @@ const converters = {
...
+        cluster: 'aqaraOpple',
+        type: ['attributeReport', 'readResponse'],
+        convert: (model, msg, publish, options, meta) => {
+            if (msg.data.hasOwnProperty(0x000A)) {
+                const lookup = {1: 'toggle', 2: 'momentary'};
+                return {switchtype: lookup[msg.data[0x000A]]};
```
Comment (Koenkk) [inline converters/fromZigbee.js:4379]: Rename this to `switch_type`.

File: devices.js [hunk 4]
```diff
@@ -1142,9 +1142,9 @@ const devices = [
         model: 'SSM-U01',
         vendor: 'Xiaomi',
         description: 'Aqara single switch module T1 (with neutral)',
-        fromZigbee: [fz.on_off, fz.metering, fz.electrical_measurement, fz.device_temperature],
-        exposes: [e.switch(), e.energy(), e.power(), e.device_temperature()],
-        toZigbee: [tz.on_off],
+        fromZigbee: [fz.on_off, fz.metering, fz.electrical_measurement, fz.device_temperature, fz.aqara_switchtype, fz.aqara_power_outage_memory],
+        exposes: [e.switch(), e.energy(), e.power(), e.device_temperature(), exposes.enum('switchtype', exposes.access.ALL, ['toggle', 'momentary']), e.power_outage_memory()],
```
Comment (Koenkk) [inline devices.js:1146]: `exposes.enum('switchtype', exposes.access.ALL, ['toggle', 'momentary'])` can be moved to `exposes.js`. Also add a description to it: `exposes.enum('switchtype', exposes.access.ALL, ['toggle', 'momentary']).withDescription("Control the behaviour of the switch")`

Comment (dmitrydyugaev) [reply devices.js]: Moved to `exposes.js`

File: converters/fromZigbee.js [hunk 5]
```diff
@@ -4370,6 +4370,26 @@ const converters = {
...
+            }
+        },
+    },
+    xiaomi_switch_power_outage_memory: {
+        cluster: 'aqaraOpple',
+        type: ['attributeReport', 'readResponse'],
+        convert: (model, msg, publish, options, meta) => {
+            if (msg.data.hasOwnProperty(0x0201)) {
+                const lookup = {false: 'false', true: 'true'};
+                return {power_outage_memory: lookup[msg.data[0x0201]]};
```
Comment (Koenkk) [inline converters/fromZigbee.js:4389]: What does this converter return, a boolean or string? (should be booelean)

Comment (dmitrydyugaev) [reply converters/fromZigbee.js:4389]: BOOLEAN

Comment (Koenkk) [reply converters/fromZigbee.js:4389]: I don't understand it, if e.g. the value of `msg.data[0x0201]` is `false` (boolean) and you take `false` from the lookup the result is `"false"` (string)

Comment (dmitrydyugaev) [reply converters/fromZigbee.js:4389]: Sorry man, I'm not sure about it. I was copying it from another device. Basically, I need to get false or true value. Not sure how to do this.

Comment (Koenkk) [reply converters/fromZigbee.js:4389]: Can you provide the debug log when this report is recevied?

See https://www.zigbee2mqtt.io/information/debug.html on how to enable debug logging.

Comment (dmitrydyugaev) [reply converters/fromZigbee.js:4389]:
```
Debug Received Zigbee message from '0x04cf8cdf3c8e92fe', type 'attributeReport', cluster 'genDeviceTempCfg', data '{"currentTemperature":23}' from endpoint 1 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8e92fe', payload '{"device_temperature":23,"linkquality":189}'
Debug Received MQTT message on 'zigbee2mqtt/0x04cf8cdf3c8e92fe/set' with data '{"power_outage_memory":true}'
Debug Publishing 'set' 'power_outage_memory' to '0x04cf8cdf3c8e92fe'
Info MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8e92fe', payload '{"device_temperature":23,"linkquality":189,"power_outage_memory":true}'
Debug Received Zigbee message from '0x04cf8cdf3c8e92fe', type 'attributeReport', cluster 'aqaraOpple', data '{"513":1}' from endpoint 1 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8e92fe', payload '{"device_temperature":23,"linkquality":189}'
Debug Received MQTT message on 'zigbee2mqtt/0x04cf8cdf3c8e92fe/set' with data '{"power_outage_memory":false}'
Debug Publishing 'set' 'power_outage_memory' to '0x04cf8cdf3c8e92fe'
Info MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8e92fe', payload '{"device_temperature":23,"linkquality":189,"power_outage_memory":false}'
Debug Received Zigbee message from '0x04cf8cdf3c8e92fe', type 'attributeReport', cluster 'aqaraOpple', data '{"513":0}' from endpoint 1 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8e92fe', payload '{"device_temperature":23,"linkquality":189}'
Debug Received MQTT message on 'zigbee2mqtt/0x04cf8cdf3c8e92fe/set' with data '{"power_outage_memory":true}'
Debug Publishing 'set' 'power_outage_memory' to '0x04cf8cdf3c8e92fe'
Debug Received Zigbee message from '0x04cf8cdf3c8e92fe', type 'attributeReport', cluster 'genDeviceTempCfg', data '{"currentTemperature":23}' from endpoint 1 with groupID 0
```

Comment (Koenkk) [reply converters/fromZigbee.js:4389]: I've added read support now. Can you open the frontend and click the refresh icon next to power_outage_memory and provide the logging of that?

Comment (dmitrydyugaev) [reply converters/fromZigbee.js:4389]:
```
Debug Received MQTT message on 'zigbee2mqtt/0x04cf8cdf3c8e92fe/get' with data '{"switch_type":""}'
Debug Publishing get 'get' 'switch_type' to '0x04cf8cdf3c8e92fe'
Debug Received Zigbee message from '0x04cf8cdf3c8e92fe', type 'readResponse', cluster 'aqaraOpple', data '{"10":1}' from endpoint 1 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8e92fe', payload '{"device_temperature":18,"energy":0,"linkquality":186,"power":0,"state":"OFF","switch_type":"toggle"}'
Debug Received Zigbee message from '0x04cf8cdf3c8e92fe', type 'attributeReport', cluster 'genDeviceTempCfg', data '{"currentTemperature":19}' from endpoint 1 with groupID 0
Info MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8e92fe', payload '{"device_temperature":19,"energy":0,"linkquality":183,"power":0,"state":"OFF","switch_type":"toggle"}'
```

Comment (Koenkk) [reply converters/fromZigbee.js:4389]: Thanks ,updated the code, can you check if it is still ok when you do a get now?

Comment (dmitrydyugaev) [reply converters/fromZigbee.js:4389]: Thanks a lot. Tested. Working like a charm.

**Pull request merged**